File: Reigo/leetcode/Python/剑指 Offer II 001. 整数除法.py
class Solution:
    def divide(self, a: int, b: int) -> int:
        __MIN__=-2**31
        __MAX__=2**31-1
        rev=False
        if a<0:
            a=-a
            rev=not rev
        if b<0:
            b=-b
            rev=not rev
        
        #print(rev)
        ret=a//b
        #print(a,b,ret)
        if rev:
            ret=-ret
        #print(ret)
        if ret>__MAX__ or ret<__MIN__:
            return __MAX__
        return ret
        

File: Reigo/leetcode/JavaScript/剑指 Offer II 001. 整数除法.js
/**
 * @param {number} a
 * @param {number} b
 * @return {number}
 */
 var divide = function(a, b) {
    __MAX__=Math.pow(2,31)-1;
    __MIN__=-Math.pow(2,31);
    if(a==0)return 0;
    var rev=false;
    if(a<0){
        rev=!rev;
        a=-a;
    }
    if(b<0){
        rev=!rev;
        b=-b;
    }

    var ans=0;
    var cnt=30;
    while(a>=b){
        //cnt-=1
        // console.log(a,b);
        // if(!cnt){
        //     return ans;
        // }
        var c=1;
        var d=b;
        while(d+d<=a){
            d=d+d;
            c=c+c;
        }
        ans+=c;
        a-=d;
    }
    if(rev)ans=-ans;
    if(ans>__MAX__ || ans<__MIN__)return __MAX__;
    return ans;
};